src/Bicep.Core/TypeSystem/DeclaredTypeManager.cs

```diff
@@ -0,0 +1,283 @@
+using Bicep.Core.Extensions;
+using Bicep.Core.Parser;
+using Bicep.Core.SemanticModel;
+using Bicep.Core.Syntax;
+
+namespace Bicep.Core.TypeSystem
+{
+    public class DeclaredTypeManager
+    {
+        // maps syntax nodes to their declared types
+        // processed nodes found not to have a declared type will have a null value
+        private readonly IDictionary<SyntaxBase, DeclaredTypeAssignment?> declaredTypes = new Dictionary<SyntaxBase, DeclaredTypeAssignment?>();
+
+        private readonly SyntaxHierarchy hierarchy;
+        private readonly ITypeManager typeManager;
+        private readonly IResourceTypeProvider resourceTypeProvider;
+        private readonly IReadOnlyDictionary<SyntaxBase, Symbol> bindings;
+
+        public DeclaredTypeManager(SyntaxHierarchy hierarchy, ITypeManager typeManager, IResourceTypeProvider resourceTypeProvider, IReadOnlyDictionary<SyntaxBase,Symbol> bindings)
+        {
+            this.hierarchy = hierarchy;
+            this.typeManager = typeManager;
+            this.resourceTypeProvider = resourceTypeProvider;
+            this.bindings = bindings;
+        }
+
+        public DeclaredTypeAssignment? GetDeclaredTypeAssignment(SyntaxBase syntax)
+        {
+            if (this.declaredTypes.TryGetValue(syntax, out var assignment))
+            {
+                // syntax node has already been processed
+                // return result as-is (even if null)
+                return assignment;
+            }
+
+            // the node has not been processed
+            // figure out the type
+            var newAssignment = GetTypeAssignment(syntax);
+
+            // cache the result
+            this.declaredTypes[syntax] = newAssignment;
+
+            return newAssignment;
+        }
+
+        public TypeSymbol? GetDeclaredType(SyntaxBase syntax) => this.GetDeclaredTypeAssignment(syntax)?.Reference.Type;
+
+        private DeclaredTypeAssignment? GetTypeAssignment(SyntaxBase syntax)
+        {
+            switch (syntax)
+            {
+                case ParameterDeclarationSyntax parameter:
+                    return GetParameterType(parameter);
+
+                case ResourceDeclarationSyntax resource:
+                    return GetResourceType(resource);
+
+                case ModuleDeclarationSyntax module:
+                    return GetModuleType(module);
+
+                case ArraySyntax array:
+                    return GetArrayType(array);
+
+                case ArrayItemSyntax arrayItem:
+                    return GetArrayItemType(arrayItem);
+
+                case ObjectSyntax @object:
+                    return GetObjectType(@object);
+
+                case ObjectPropertySyntax objectProperty:
+                    return GetObjectPropertyType(objectProperty);
+            }
+
+            return null;
+        }
+
+        private DeclaredTypeAssignment GetParameterType(ParameterDeclarationSyntax syntax) => new DeclaredTypeAssignment(syntax.GetDeclaredType());
```

Comment:
Other than renames and minor tweaks to get things to compile, the logic in all the methods below hasn't changed.